pkg/pillar/dpcreconciler/linux.go

```diff
@@ -687,6 +707,81 @@ func (r *LinuxDpcReconciler) clusterStatusChanged(
 			newStatus.ClusterIPPrefix)
 }
 
+// kubeUserServicesChanged checks if the Kubernetes user services configuration has changed
+func (r *LinuxDpcReconciler) kubeUserServicesChanged(
+	newServices types.KubeUserServices) bool {
+	// Check if service lists are different
+	if len(r.lastArgs.KubeUserServices.UserService) != len(newServices.UserService) {
+		return true
+	}
+
+	// Check if ingress lists are different
+	if len(r.lastArgs.KubeUserServices.UserIngress) != len(newServices.UserIngress) {
+		return true
+	}
+
+	// Create maps for fast lookup of existing services
+	oldServiceMap := make(map[string]types.KubeServiceInfo)
+	for _, svc := range r.lastArgs.KubeUserServices.UserService {
+		key := svc.Namespace + "/" + svc.Name
+		oldServiceMap[key] = svc
+	}
+
+	// Check if any service has been added, removed or modified
+	for _, newSvc := range newServices.UserService {
+		key := newSvc.Namespace + "/" + newSvc.Name
+		oldSvc, exists := oldServiceMap[key]
+		if !exists ||
+			oldSvc.Protocol != newSvc.Protocol ||
+			oldSvc.Port != newSvc.Port ||
+			oldSvc.NodePort != newSvc.NodePort ||
+			oldSvc.LoadBalancerIP != newSvc.LoadBalancerIP ||
+			oldSvc.Type != newSvc.Type {
+			return true
+		}
+	}
+
+	// Create maps for fast lookup of existing ingresses
+	oldIngressMap := make(map[string]types.KubeIngressInfo)
+	for _, ing := range r.lastArgs.KubeUserServices.UserIngress {
+		key := ing.Namespace + "/" + ing.Name + "/" + ing.Hostname + ing.Path
+		oldIngressMap[key] = ing
+	}
+
+	// Check if any ingress has been added, removed or modified
+	for _, newIng := range newServices.UserIngress {
+		key := newIng.Namespace + "/" + newIng.Name + "/" + newIng.Hostname + newIng.Path
+		oldIng, exists := oldIngressMap[key]
+		if !exists ||
+			oldIng.PathType != newIng.PathType ||
+			oldIng.ServiceType != newIng.ServiceType ||
+			oldIng.Protocol != newIng.Protocol ||
+			!compareIngressIPs(oldIng.IngressIP, newIng.IngressIP) {
```

Comment:
You can also just call `generics.EqualSets` for list of strings.

Reply:
Ok, will do.